src/lib.rs

```diff
@@ -67,6 +78,8 @@ pub trait DfuIo {
     type Reset;
     /// Error type.
     type Error: From<Error>;
+    /// Dfuse Memory layout type
+    type MemoryLayout: AsRef<memory_layout::mem>;
```

Comment:
I'm not sure why you needed a type argument for this?

Reply:
To provide the type for `DfuProtocol<M>`   which used as the return type for the protocol() function;  Otherwise we'd need to pull it up as a trait generic parameter which would also leak out to DfuSansIO, which makes it more annoying to use

Reply:
I guess I expressed myself wrong. Why do we need a generic over memory layout? Can't we just use the type directly?

Reply:
Ah because that allows you to to just have your trait implement contain a `DfuProtocol<MemoryLayout>` and return a reference to that in the trait which is quite convenient; But ofcourse `MemoryLayout` can't be used on a no_std environment so we can't just use that.

returning `&DfuProtocol<&memory_layout::mem>` (or a non-generic DFuProtocol with that type internally) would be tricky as that will quickly run into self-referential types if you want to save the DfuProtocol in your struct. 

The last option would be to return an owned `DfuProtocol<&memory_layout::mem>` but then you have to create that in the trait method, probably converted from a `DfuProtocol<Memory:Layout>` on `std` systems on demand.. Which also isn't really any nicer for the implementer

So the type parameter seemed the most convenient way to go about it ; but maybe i overlooked something